DESIGN.md

```diff
@@ -0,0 +1,161 @@
+Currently the first option is used.
+This has the advantage that fewer objects are passing through the channels.
+It also works better with the current map-reduce functions in core sense they expect all the data in a single object.
+It has the disadvantage that the whole batch has to be held in memory.
+In some cases the entire batch does need to live in memory but not in all.
+For example a node that is counting points per batch need only maintain a counter in memory and not the entire batch.
+
+### Source Mapping
+
+Kapacitor can receive data from many different sources, including querying InfluxDB.
+The type TaskMaster in`task_master.go` is responsible for managing which tasks are receiving which data.
+
+For stream tasks this is done by having on global edge.
+All sources (graphite, collectd, http, etc) write their data to the TaskMaster, who writes the data to the global *stream* edge.
+When a stream task is started it gets a *fork* of the global stream filtered down by the databases and retention policies its allowed to access.
+Then the task can further process the data stream.
+
+In the case of the batch tasks, the TaskMaster manages starting the schedules for querying InfluxDB.
+The results of the queries are passed to the root nodes of the task directly.
+
+
+### Windowing
+
+Windowing data is an important piece to creating pipelines.
+Windowing is concerned with how you can slice a data stream into multiple windows and is orthogonal to how batches are transfered.
+Kapacitor handles windowing explicitly, by allowing the user to define a WindowNode
+that has two parameters. First, the `period` is the length of the window in time.
+Second, the `every` property defines how often an window should be emitted into the stream.
+This allows for creating windows that overlap, have no overlap, or have gaps between the windows.
+As a result the concept of a window does not exist inherently in the data stream, but rather windowing is the method of converting a stream of data into a batch of data.
+
+Example TICKscript:
+
+```javascript
+stream
+    .window()
+        .period(10s)
+        .every(5s)
+```
+
+The above script slices the incoming stream into overlapping windows.
+Each window contains the last 10s of data and a new window is emitted every 5s.
+
+
+### Challenges
+
+Challenges with the current implementation:
+
+* For stream tasks: If a single node stop processing data all nodes will eventually stop including nodes from other tasks.
+    This is because of the global stream to aggregate all incoming sources and the fact the edges just block instead of dropping data.
+    This could be mitigated further by creating independent streams for each database retention policy pair, but this only provides isolation and not a solution.
+    We need a contract in place for what to do when a given node stops processing data.
+* Nodes are responsible for not creating deadlock in the way they read and write data from their parent and child edges.
```

Comment:
Shim multiplexing above join node instead of inside of it. Single input and output per node